zerver/tornado/handlers.py

```diff
@@ -339,16 +155,90 @@ def zulip_finish(self, result_dict: Dict[str, Any], request: HttpRequest,
-        # the headers from that since sending those to Tornado seems
-        # tricky; instead just send the (already json-rendered)
-        # content on to Tornado
-        django_response = json_response(res_type=result_dict['result'],
-                                        data=result_dict, status=self.get_status())
-        django_response = self.apply_response_middleware(request, django_response,
-                                                         request._resolver)
+        # The `result` dictionary contains the data we want to return
+        # to the client.  We want to do so in a proper HTTP response
+        # after running our Django response middleware (which does
+        # things like log the request, add rate-limit headers, etc.).
+        # Making an HttpResponse object with json_response is easy.
+        # What's tricky is getting Django to run process_response
+        # middleware on it without writing a lot of duplicate code.
+        #
+        # Django runs the process_response middleware functions inside
+        # MiddlewareMixin.__call__, which also runs the
+        # process_request middleware.  That code path is activated
+        # inside Django BaseHandler's `self.get_response` and
+        # `load_middleware` functions, which effectively end up doing:
+        #    response = m1(m2(m3(...(self._get_response(request)))))
+        #
+        # where the m1 (etc.) methods are the middleware classes
+        # (i.e. their __call__) methods.  We've already run that code
+        # once when we received the request (we discarded the
+        # resulting HttpResponse in order to long-poll).
+        #
+        # So at this point, we just want to run the response
+        # middleware (the request middleware was already run before we
+        # chose to long-poll this request by not returning an
+        # HttpResponse), we:
+        #
+        # * Generate an HttpResponse object as a Zulip view function
+        #   would use json_response.
+        #
+        # * Store that HttpResponse object in self.saved_response,
+        #   configuring our _get_response wrapper to return it. (We
+        #   could instead have just written the actual view code to
+        #   return the response we want, but that would result in
+        #   _get_response running Django's process_view middleware on
+        #   the HttpRequest a second time, which could result in weird
+        #   bugs).
+        #
+        # * By setting request.request_middleware_run, we configure
+        #   the below middleware_mixin__call___hack monkey-patching to
+        #   arrange the process_request middleware being skipped to
+        #   avoid running it twice, and instead only the
+        #   process_response middleware being run.
+        #
+        # * Then we just call `self.get_response`, passing our saved
+        #   HttpRequest, and send the resulting content to the client.
+
+        request.request_middleware_run = True
```

Comment:
This variable name looked wrong until I realized you intended “middleware [was] run” rather than “middleware [should] run”. Perhaps `ran_request_middleware` or `skip_request_middleware` would be clearer.

Reply:
I like "skip" -- that feels clearer to me both here and in the implementation, below.